src/runtime-rs/crates/hypervisor/src/device/topology.rs

```diff
@@ -203,14 +203,143 @@ pub struct PCIeRootComplex {
+}
+
+impl std::fmt::Display for PCIePortBusPrefix {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let prefix = match self {
+            PCIePortBusPrefix::RootPort => "rp",
+            PCIePortBusPrefix::SwitchPort => "sw",
+            PCIePortBusPrefix::SwitchUpstreamPort => "swup",
+            PCIePortBusPrefix::SwitchDownstreamPort => "swdp",
+        };
+        write!(f, "{}", prefix)
+    }
+}
+
+/// PCIePort distinguishes between different types of PCIe ports.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Hash)]
+pub enum PCIePort {
+    // NoPort is for disabling VFIO hotplug/coldplug
+    #[default]
+    NoPort,
+
+    /// RootPort attach VFIO devices to a root-port
+    RootPort,
+
+    // SwitchPort attach VFIO devices to a switch-port
+    SwitchPort,
+
+    // InvalidPort is for invalid port
+    InvalidPort,
+}
+
+impl std::fmt::Display for PCIePort {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let port = match self {
+            PCIePort::NoPort => "no-port",
+            PCIePort::RootPort => "root-port",
+            PCIePort::SwitchPort => "switch-port",
+            PCIePort::InvalidPort => "invalid-port",
+        };
+        write!(f, "{}", port)
+    }
+}
+
+/// Represents an available node in the PCIe topology.
+#[derive(Clone, Debug)]
+pub enum AvailabledNode {
+    TopologyPortDevice(TopologyPortDevice),
+    SwitchDownPort(SwitchDownPort),
+}
+
+/// Represents a PCIe port
+#[derive(Default, Clone, Debug)]
+pub struct SwitchDownPort {
+    pub id: u32,
+    pub bus: String, // swupX
+    pub allocated: bool,
+    pub connected_device: Option<String>,
+}
+
+impl SwitchDownPort {
+    pub fn bus(&self) -> String {
+        format!("{}{}", PCIePortBusPrefix::SwitchDownstreamPort, self.id)
+    }
+}
+
+/// Represents a PCIe switch
+#[derive(Debug, Clone)]
+pub struct PcieSwitch {
+    pub id: u32,
+    pub bus: String,                                //rpX
+    pub switch_ports: HashMap<u32, SwitchDownPort>, // Switch ports
+}
+
+impl PcieSwitch {
+    pub fn bus(&self) -> String {
+        format!("{}{}", PCIePortBusPrefix::SwitchUpstreamPort, self.id)
+    }
+}
+
+/// Represents a root port attached on root bus, TopologyPortDevice represents a PCIe device used for hotplugging.
+#[derive(Debug, Clone)]
+pub struct TopologyPortDevice {
+    pub id: u32,
+    pub bus: String, //pcie.0
+    pub allocated: bool,
+    pub connected_switch: Option<PcieSwitch>, // Connected PCIe switch
+}
+
+impl TopologyPortDevice {
+    pub fn new(id: u32, bus: &str) -> Self {
+        Self {
+            id,
+            bus: bus.to_owned(),
+            allocated: false,
+            connected_switch: None,
+        }
+    }
+
+    pub fn get_port_type(&self) -> PCIePort {
+        match self.connected_switch {
+            Some(_) => PCIePort::SwitchPort,
+            None => PCIePort::RootPort,
+        }
+    }
+
+    pub fn bus(&self) -> String {
```

Comment:
Maybe it's just me but I find it a bit confusing to have both `TopologyPortDevice::bus` and `TopologyPortDevice::bus()`, meaning different things.  Could one of them be better called something else?

Reply:
Sorry here, It's indeed misleading and I correct it with `port_id`, does it make sense ?

Reply:
I intentionally refrained from making a specific suggestion since I'm not familiar enought with PCIe terminology but `port_id` makes a lot of sense to me.  Thanks!